### PR TITLE
Fix compilation when forcing later C++.

### PR DESCRIPTION
**Pull request by vrabaud**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch

### CODE REVIEW

**Comment by vrabaud** [comment]
The CI on OSX failed because of this: https://github.com/opencv/opencv/actions/runs/5964752904/job/16180691776?pr=24193

File: modules/objdetect/test/test_qrcode_encode.cpp [hunk 1]
```diff
@@ -33,6 +33,18 @@ std::string encode_qrcode_eci_images_name[] = {
 const Size fixed_size = Size(200, 200);
 const float border_width = 2.0;
 
+void randomShuffle(std::string& info) {
+#if defined CV_CXX11
```
Comment (mshabunin) [inline modules/objdetect/test/test_qrcode_encode.cpp:37]: I thought we are going to remove `CV_CXX11` completely, see #23736
**cc** @asmorkalov , @opencv-alalek , @vpisarev

Comment (vrabaud) [reply modules/objdetect/test/test_qrcode_encode.cpp:37]: Indeed. Plus, OpenCV 4.x is C++11 so we can just use std::shuffle by default.

**Review by mshabunin — APPROVED** [review]
Reproduced with libstdc++ (warning) and libc++ (error) and clang 17 (`CMAKE_CXX_STANDARD=17`):
```
/opencv/modules/objdetect/test/test_qrcode_encode.cpp:267:22: warning: 'random_shuffle<__gnu_cxx::__normal_iterator<char *, std::basic_string<char>>>' is deprecated: use 'std::shuffle' instead [-Wdeprecated-declarations]
  267 |                 std::random_shuffle(input_info.begin(),input_info.end());
      |                      ^


/opencv/modules/objdetect/test/test_qrcode_encode.cpp:267:22: error: no member named 'random_shuffle' in namespace 'std'
  267 |                 std::random_shuffle(input_info.begin(),input_info.end());
      |                 ~~~~~^
```
No issues after fix.

**Pull request merged**